src/cfg/cfg-traversal.h

```diff
@@ -210,41 +219,89 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
     // unless the call is within a try-catch, because the CFG will have too many
     // blocks that way, and if an exception is thrown, the function will be
     // exited anyway.
-    if (!self->catchStack.empty()) {
+    if (!self->unwindCatchStack.empty()) {
       auto* last = self->currBasicBlock;
       self->startBasicBlock();
       self->link(last, self->currBasicBlock);    // exception not thrown
-      self->link(last, self->catchStack.back()); // exception thrown
+      // Exception thrown. Create a link to each catch within the innermost try.
+      for (auto* block : self->unwindCatchStack.back()) {
+        self->link(last, block);
+      }
+      // If the innermost try does not have a catch_all clause, an exception
+      // thrown can be caught by any of its outer catch block. And if that outer
+      // try-catch also does not have a catch_all, this continues until we
+      // encounter try-catch_all. Create a link to all those possible catch
+      // unwind destinations.
+      for (int i = self->unwindCatchStack.size() - 1; i > 0; i--) {
+        if (self->unwindExprStack[i]->template cast<Try>()->hasCatchAll()) {
+          break;
+        }
+        for (auto* block : self->unwindCatchStack[i - 1]) {
+          self->link(last, block);
+        }
+      }
     }
   }
 
   static void doStartTry(SubType* self, Expression** currp) {
+    auto* curr = (*currp)->cast<Try>();
     auto* last = self->currBasicBlock;
-    self->startBasicBlock(); // catch body's first block
-    self->catchStack.push_back(self->currBasicBlock);
+    self->unwindCatchStack.emplace_back();
+    self->unwindExprStack.push_back(curr);
+    for (Index i = 0; i < curr->catchBodies.size(); i++) {
+      self->startBasicBlock(); // catch body's first block
+      self->unwindCatchStack.back().push_back(self->currBasicBlock);
```

Comment:
these two lines might be clearer maybe if `startBasicBlock` returned what it created? then this could all be done on one line.

Reply:
Done. Fix other uses of `startBasicBlock` after which we only have a single use of `self->currBasicBlock`.